pyccel/parser/syntax/openmp.py

```diff
@@ -131,60 +230,504 @@ def expr(self):
         _valid_clauses = (OmpPrivate, \
                          OmpFirstPrivate)
 
-        txt = 'single'
-        for clause in self.clauses:
-            if isinstance(clause, _valid_clauses):
-                txt = '{0} {1}'.format(txt, clause.expr)
-            else:
-                raise TypeError('Wrong clause for OmpSingleConstruct')
+        txt = self.name
+        for clause in self.clauses:
+            if isinstance(clause, _valid_clauses):
+                txt = '{0} {1}'.format(txt, clause.expr)
+            else:
+                raise TypeError('Wrong clause for OmpSingleConstruct')
+
+        return OMP_Single_Construct(txt)
+
+class OmpCriticalConstruct(BasicStmt):
+    """Class representing a Critical construct."""
+    def __init__(self, **kwargs):
+        """
+        """
+        self.clauses = kwargs.pop('clauses')
+        self.name    = kwargs.pop('name')
+
+        super().__init__(**kwargs)
+
+    @property
+    def expr(self):
+        if DEBUG:
+            print("> OmpCriticalConstruct: expr")
+
+        _valid_clauses = (OmpCriticalName)
+
+        txt = self.name
+        for clause in self.clauses:
+            if isinstance(clause, _valid_clauses):
+                txt = '{0} {1}'.format(txt, clause.expr)
+            else:
+                raise TypeError('Wrong clause for OmpCriticalConstruct')
```

Comment:
Several of these classes contain these same lines of code. Could these be defined as a function of a common superclass to reduce code repetition?

Reply:
See `PyccelOperator` for an example of using super classes to reduce code duplication as much as possible